### ISSUE TITLE
Canvas API

### ISSUE THREAD

**Issue by maximecb** [maximecb/zeta#1]
Zeta will not have an FFI, at least not in the beginning. Instead, it will expose several low-level APIs as zeta runtime functions. APIs for things such as 2D graphics, file access, networking, etc. The APIs themselves will be implemented in C. High-level libraries will be built on top of these APIs, and will be written in the zeta language.

One of the first APIs I would like to create is for a canvas window, using X11. This API will be very minimalistic. It will allow plotting individual pixels, drawing lines, filling rectangles and also blitting bitmaps. That is all, no other functionality will be provided. The reason why I want to keep things simple is so that the API remains unlikely to break, and so we can guarantee that graphics drawn using the said API look the same on every system, even if we were to replace the X11 implementation by one written using a different library. Functions for keyboard and mouse input should be added at a later stage to allow creating interactive applications.

The first thing to do will be to write C code to interface with X11, create a window, and plot pixels/lines/etc. Then, these functions will be exposed to zeta code through wrappers. A canvas window zeta library can then be written on top of these. This will be a different issue.

**Comment by bFraley** [comment]
The C API  can be a simple header file that will include the X11 headers.
There, structs and functions wrapping the X11 values of a new window (like i've done with window_init() in the example below) and then some routines that call and execute draw functions from a program source (much like the callback functions in Higgs), can be defined.

The window_init() function is how we can group and initialize the series of x11 calls into a single function callable from a zeta.  Except here, there is the luxury of directly calling X11 functions - Note: We can use pointers directly to X11 functions and our own wrapper functions, if needed.

As you'll see in the example, we just need the display, screen, gc (graphic context), and black and white pixel variables to get started. The values for these come from directly calling into their functions - populating their associated structure and values. Once a window is created, we need to listen to the ExposureMask in order for calls like XDrawLine to render. But it is as simple as that - we can directly call XDrawLine, XDrawRectangle, etc. We just need a few functions to pass in the parameters from the higher level zeta expressions.

It won't take much to add support for colors, keys, pixmaps, etc..
Basically, there isn't much 'glue' needed, especially if a zeta command directly points to our API routines and X11 functions...I'm not sure how complex this may be - or represented in the context of the VM / JIT and program heap..? But it should be as simple as what I've done here below.

https://gist.github.com/bFraley/11f60f20c42211bbdf70

**Comment by bFraley** [comment]
Additionally, for the X11 API and others like Issue #4, I imagine the design and workflow is to:
-Create a header file that groups (wraps) calls to shared lib functions.
-Directly point to / call these functions from zeta.

It will  be a matter of understanding the libs / APIs we want to use, being careful with types and function parameter sizes, and only exposing what we really need.

**Comment by maximecb** [comment]
I don't want to expose X11 functions or structs directly to zeta. I want to keep things abstract so we can make the API as simple and portable as possible.

The wrappers will be zeta objects containing a C function pointer and meta data such as the expected number of arguments.
